backend/cfg/vectorize.ml

```diff
@@ -178,61 +315,501 @@ end = struct
+       certain that y does not depend on that fresh allocation, then they point
+       to disjoint addresses *)
+    (* At each load or store instruction, it keeps track of all allocs up to
+       this point in this basic block and puts them in one of 3 categories:
+       [fresh_allocs]: nothing that depends on the address of the fresh alloc
+       has been saved as a value; [stored_allocs]: something that depends on the
+       address of the fresh alloc has been saved as a value, but nothing has
+       been loaded till this point; [unsure_allocs]: something that depends on
+       the address of the fresh alloc has been saved as a value, and something
+       has been loaded till this point. For each memory operation, we will save
+       its dependent allocs and unsure allocs *)
+    let dependency_graph = Dependency_graph.from_block block in
+    let id_to_instructions =
+      DLL.to_list block.body
+      |> List.map (fun basic_instruction ->
+             let instruction = Instruction.Basic basic_instruction in
+             Instruction.id instruction, instruction)
+      |> Instruction.Id.Tbl.of_list
+    in
+    let memory_operations = Instruction.Id.Tbl.create 100 in
+    let loads, stores, _, _, _ =
+      DLL.fold_left block.body
+        ~f:
+          (fun (loads, stores, fresh_allocs, stored_allocs, unsure_allocs)
+               basic_instruction ->
+          let instruction = Instruction.Basic basic_instruction in
+          let id = Instruction.id instruction in
+          if Instruction.is_alloc instruction
+          then
+            ( loads,
+              stores,
+              Instruction.Id.Set.add id fresh_allocs,
+              stored_allocs,
+              unsure_allocs )
+          else
+            let memory_operation = Memory_operation.init instruction in
+            match memory_operation with
+            | None ->
+              if Instruction.may_break_alloc_freshness instruction
+              then
+                ( loads,
+                  stores,
+                  Instruction.Id.Set.empty,
+                  Instruction.Id.Set.empty,
+                  Instruction.Id.Set.union fresh_allocs stored_allocs
+                  |> Instruction.Id.Set.union unsure_allocs )
+              else loads, stores, fresh_allocs, stored_allocs, unsure_allocs
+            | Some memory_operation -> (
+              let get_dependent_allocs_of_arg arg_i =
+                Dependency_graph.get_all_dependencies_of_arg dependency_graph id
+                  ~arg_i
+                |> Instruction.Id.Set.filter
+                     (Instruction.is_alloc
+                     << Instruction.Id.Tbl.find id_to_instructions)
+              in
+              let start_index =
+                match memory_operation.op with Load -> 0 | Store -> 1
+              in
+              let dependent_allocs, _ =
+                Array.fold_left
+                  (fun (dependent_allocs, arg_i) _ ->
```

Comment:
This `fold_left` can be simplified, if the argument is unused. Also `start_index` should be abstracted into `Memory_operation`, similarly to the way `memory_arguments` is.

Reply:
done